lib/strings.nix

```
@@ -798,10 +799,31 @@ rec {
   in lib.warnIf (!precise) "Imprecise conversion from float to string ${result}"
     result;
 
-  /* Check whether a value can be coerced to a string */
-  isCoercibleToString = x:
-    elem (typeOf x) [ "path" "string" "null" "int" "float" "bool" ] ||
-    (isList x && lib.all isCoercibleToString x) ||
+  /* Soft-deprecated function. While the original implementation is available as
+     isConvertibleWithToString, consider using isStringLike instead, if suitable. */
+  isCoercibleToString = lib.warnIf (lib.isInOldestRelease 2305)
+    "lib.strings.isCoercibleToString is deprecated in favor of either isStringLike or isConvertibleWithToString. Only use the latter if it needs to return true for null, numbers, booleans and list of similarly coercibles."
+    isConvertibleWithToString;
+
+  /* Check whether a list or other value can be passed to toString.
+
+     Many types of value are coercible to string this way, including int, float,
+     null, bool, list of similarly coercible values.
+  */
+  isConvertibleWithToString = x:
+    isStringLike x ||
+    elem (typeOf x) [ "null" "int" "float" "bool" ] ||
+    (isList x && lib.all isConvertibleWithToString x);
+
+  /* Check whether a value can be coerced to a string.
+     The value must be a string, path, or attribute set.
+
+     StringLike values can be used without explicit conversion in most functions
+     that expect a string.
```

```suggestion
     String-like values can be used without explicit conversion in most places
     that expect a string.
```

Importantly, in interpolations.

I've actually had in my TODO list for a while to add a `types.stringLike` meant to replace the `either str path` (or, worse, `either str package`) idiom e.g. in [`system.build.installBootLoader`](https://github.com/NixOS/nixpkgs/blob/83d75023879180ffe21169216ad86573a153a303/nixos/modules/system/activation/top-level.nix#L188), [`console.font`](https://github.com/NixOS/nixpkgs/blob/83d75023879180ffe21169216ad86573a153a303/nixos/modules/config/console.nix#L48), etc. Essentially options whose type is "anything Nix accepts in an interpolation".

---

Applied.



> [`console.font`](https://github.com/NixOS/nixpkgs/blob/83d75023879180ffe21169216ad86573a153a303/nixos/modules/config/console.nix#L48),

I think the current type for this one is actually nice, because the two types have quite different interpretations. It adds a bit of meaning compared to `stringLike`, even if that would be sufficient.